llvm/include/llvm/CodeGen/SelectionDAGNodes.h

```diff
@@ -708,15 +708,7 @@ END_TWO_BYTE_PACK()
   bool isUndef() const { return NodeType == ISD::UNDEF; }
 
   /// Test if this node is a memory intrinsic (with valid pointer information).
-  /// INTRINSIC_W_CHAIN and INTRINSIC_VOID nodes are sometimes created for
-  /// non-memory intrinsics (with chains) that are not really instances of
-  /// MemSDNode. For such nodes, we need some extra state to determine the
-  /// proper classof relationship.
-  bool isMemIntrinsic() const {
-    return (NodeType == ISD::INTRINSIC_W_CHAIN ||
-            NodeType == ISD::INTRINSIC_VOID) &&
-           SDNodeBits.IsMemIntrinsic;
-  }
+  bool isMemIntrinsic() const { return SDNodeBits.IsMemIntrinsic; }
```

Comment:
Isn't this SDNodeBits part of a union? Isn't the opcode check to make sure that part of the union is active?

Reply:
structs other than `SDNodeBitfields` have first `NumSDNodeBits` reserved.
That is, `IsMemIntrinsic` is shared between all structs in the union.

I think the opcode check is a historical artefact.

Reply:
Note that this opcode check doesn't really tell us which union member is active because `INTRINSIC_W_CHAIN` / `INTRINSIC_VOID` can either be an ordinary `SDNode` or `MemIntrinsicSDNode`. If I read the removed comment correctly, this bit was introduced to disambiguate these cases, but introducing it also made the opcode check redundant.

Assuming there is no (pre-existing) undefined behavior, just checking the bit is more robust as the method will now return false for ordinary `SDNode`s erroneously created with `isTargetMemoryOpcode` opcodes.

Reply:
Thanks for the clarification.

> Assuming there is no (pre-existing) undefined behavior, just checking the bit is more robust as the method will now return false for ordinary SDNodes erroneously created with isTargetMemoryOpcode opcodes.

I think that would be considered a bug.